cmake/grpc.cmake

```diff
@@ -11,39 +11,115 @@ unset(_grpc_cxx_standard)
 unset(_package_providers)
 unset(_patch_clause)
 
-function(prepare_grpc_patch_step OUTVAR)
-  if(NOT DEFINED GRPC_VERSION OR GRPC_VERSION STREQUAL "")
-    message(WARNING "GRPC_VERSION undefined; cannot select patch file")
+#-----------------------------------------------------------------------
+# define_grpc_patch_file()
+# Returns the filename of the template patch file.
+#
+# Parameters:
+#   PATCHDIR - directory containing patch files.
+#   OUTVAR - name of output variable.
+#
+# Variables:
+#   GRPC_PATCH - name of gRPC patch file.
+#   GRPC_VERSION - version of gRPC being patched.
+#
+# Choices are, in descending order of precedence:
+#  1. ${GRPC_PATCHFILE}
+#  2. grpc-v$(GRPC_VERSION}.patch.in
+#  3. grpc-current.patch.in
+#
+# Returns an empty string if we cannot determine a valid patch file name.
+#-----------------------------------------------------------------------
+function(define_grpc_patch_file PATCHDIR OUTVAR)
+  # Specified patch file
+  if(DEFINED GRPC_PATCHFILE AND NOT GRPC_PATCHFILE STREQUAL "")
+    set(patchfile "${GRPC_PATCHFILE}")
+    if(EXISTS ${PATCHDIR}/${patchfile})
```

Comment:
Is the user allowed to specify a a ${patchfile} as a build parameter (not sure if that is the intention based on the choice no 1 in comment above)? If yes, adding a ${PATCHDIR} to a provided absolute path will not work